Read message bundles in UTF-8 and fall back to 8859-1 in `org.eclipse.osgi.util.NLS`

ShadelessFox opened the pull request:
All message bundles are currently loaded in the ISO 8859-1 character encoding. This enforces any message bundle that uses non-ASCII characters to escape such characters using escape sequences. This makes it notoriously hard to analyze or process such files without using specialized tools such as IDEs or relying on the `native2ascii` command-line tool.

This PR makes UTF-8 the default encoding when using `org.eclipse.osgi.util.NLS#initializeMessages` by leveraging the usage of [PropertyResourceBundle](https://docs.oracle.com/javase/8/docs/api/java/util/PropertyResourceBundle.html). Despite interpreting the stream in UTF-8 by default, it still correctly handles 8859-1 that is not compatible with UTF-8 by re-reading the stream from the start if it cannot be properly decoded in UTF-8 otherwise.

vogella commented:
Please remove merge commit.

merks commented:
FYI, we typically expect folks to follow these steps:

https://github.com/eclipse-simrel/.github/pull/34

So typically folks will amend and force push to maintain a single commit in the PR...

ShadelessFox commented:
I pulled changes from the upstream and removed the unwanted merge commit.

merks commented:
The build fails with:

> 06:10:29  [ERROR] Failed to execute goal org.eclipse.tycho.extras:tycho-p2-extras-plugin:4.0.10:compare-version-with-baselines (compare-attached-artifacts-with-release) on project org.eclipse.equinox.supplement: Only qualifier changed for (org.eclipse.equinox.supplement/1.11.100.v20241203-1040). Expected to have bigger x.y.z than what is available in baseline (1.11.100.v20241030-2121) -> [Help 1]

I think this is new API (right @tjwatson ?) so there needs to be an `@since` and this bundle needs to be incremented to 1.12.0

![image](https://github.com/user-attachments/assets/6ca14f08-0ef2-4160-94fa-f80ce7ab9b6a)

ShadelessFox commented:
Since there's no `package-info` in the enclosing package, can @laeubi's suggestion be safely omitted?

laeubi commented on bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java:
```diff
@@ -141,18 +145,50 @@ public static String bind(String message, Object[] bindings) {
+	 * @param clazz the class where the constants will exist
+	 */
+	public static void initializeMessages(final String baseName, final Class<?> clazz) {
+		initializeMessages(baseName, clazz, StandardCharsets.UTF_8);
```
I'm not sure UTF-8 is a good default here, as it will most likely break a lot of exiting behavior.

ShadelessFox replied:
This PR's whole purpose is to make UTF-8 the default. I tested it locally against various languages; the behavior is the same against both ASCII-only messages and messages containing escape sequences (this includes anything outside the ASCII range). Please see my original comment on this PR.

laeubi replied:
Please test with a file encoded with `ISO_8859_1` special chars, ASCII is not the default and escaping is not mandatory see [here](https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html)

> The [load(InputStream)](https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#load-java.io.InputStream-) / [store(OutputStream, String)](https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#store-java.io.OutputStream-java.lang.String-) methods work the same way as the load(Reader)/store(Writer, String) pair, except the input/output stream is **encoded in ISO 8859-1 character encoding**. Characters that **cannot be directly represented in this encoding** can be written using Unicode escapes as defined in section 3.3 of The Java™ Language Specification;

This means that it is totally valid to have any ISO 8859-1 in such a file and reading it as UTF-8 will break these existing file.

merks replied:
Yes, one must assume that ISO 8859-1 is commonly used.   it seems to me that Java was making some changes in this area and I found this:

https://docs.oracle.com/javase/9/intl/internationalization-enhancements-jdk-9.htm#JSINT-GUID-5ED91AA9-B2E3-4E05-8E99-6A009D2B36AF

![image](https://github.com/user-attachments/assets/368d0e56-ca26-4b49-9b63-608794a753f1)

It's definitely a bad idea to assume UTF-8.

laeubi commented:
> Since there's no `package-info` in the enclosing package, can @laeubi's suggestion be safely omitted?

This is completely independent from `package-info` ...

ShadelessFox commented:
I'm a bit lost. Can you please hint where that package version you're talking about is located?

laeubi commented:
> > This is completely independent from package-info ...
> 
> I'm a bit lost. Can you please hint where that package version you're talking about is located?

https://github.com/eclipse-equinox/equinox/blob/fd8cb34886d7fcd09753c6bd0da61c31a8f9697e/bundles/org.eclipse.osgi/supplement/META-INF/MANIFEST.MF#L20

laeubi commented on bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java:
```diff
@@ -141,18 +145,50 @@ public static String bind(String message, Object[] bindings) {
 	 * @param baseName the base name of a fully qualified message properties file.
 	 * @param clazz the class where the constants will exist
+	 * @param charset the charset for the given properties file
+	 * @since 3.22
```
```suggestion
	 * @since 1.12
```

ShadelessFox replied:
Yes, that makes more sense. Eclipse complained I should use `3.22` instead. Noted.

ShadelessFox commented:
Updated the PR while hopefully addressing all issues. Please let me know if I missed something.

laeubi commented:
By the way I think the best way forward would be to add support for [UTF-8 BOM](https://en.wikipedia.org/wiki/Byte_order_mark) together with a buffered stream of 4 bytes  this would allow to detect if the file is "standard" or UTF encoded.

laeubi reviewed (changes requested):
I don't think it works in the current form an `ISO_8859_1` must be the default, beside that one should add at least some test-cases to ensure the code works as expected as this is a very central class used everywhere in Eclipse there is a high risk of breaking things (now or in the future) without at least minimal test coverage..

ShadelessFox commented:
Thanks for the feedback. Hopefully, I'll get my hands on this in the future. :^)

ShadelessFox commented:
> By the way I think the best way forward would be to add support for [UTF-8 BOM](https://en.wikipedia.org/wiki/Byte_order_mark) together with a buffered stream of 4 bytes this would allow to detect if the file is "standard" or UTF encoded.

Unfortunately, the BOM's byte values are valid ISO 8859-1 characters unless we assume no one is ever going to use `ï»¿` as the first three characters of a message bundle.

laeubi commented:
> Unfortunately, the BOM's byte values are valid ISO 8859-1 characters unless we assume no one is ever going to use `ï»¿` as the first three characters of a message bundle.

I think we can safely assert that this is very unlikely as messages are usually mapped to Java constants and this is not a valid java identifier.

merks commented:
As I mentioned above, see this comment:

> Most existing properties files should not be affected: UTF-8 and ISO-8859-1 have the same encoding for ASCII characters, and human-readable non-ASCII ISO-8859-1 encoding is not valid UTF-8. If an invalid UTF-8 byte sequence is detected, the Java runtime automatically rereads the file in ISO-8859-1.

That sounds to me like a much better approach, especially if that's what Java itself does.

laeubi commented:
> Here too:
> 
> https://docs.oracle.com/javase/9/docs/api/java/util/PropertyResourceBundle.html
> 
> So it seem to me that we don't need new APIs. We could simply implement the same behavior.

I never understood why platform has its own mechanism instead of reusing the the java one, but there must be a reason :-)

merks commented:
It is reusing Java, using it to set the fields reflectively:

https://github.com/eclipse-equinox/equinox/blob/793dcc532e6b66cd8adb3fb70aa6fe5c03b774c6/bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java#L409-L467

tjwatson commented:
I also agree that some tests are needed.

ShadelessFox commented:
I have updated PR, removing the newly introduced API and instead relying on the BOM for UTF-8, UTF-16 LE, and UTF-16 BE. The old behavior before this PR is kept intact; if no BOM is present, the 8859-1 encoding will be used instead. 

The suggested approach of interpreting data as UTF-8 and falling back to 8859-1, as done in Java's [PropertyResourceBundle](https://docs.oracle.com/javase/9/docs/api/java/util/PropertyResourceBundle.html), seems tricky to get right and not as reliable as BOM, so I haven't considered following that path.

I have also changed the commit's name to reflect its new intention, but the branch name is now a bit misleading. Should I close this PR and open a new one instead? Also, is it worth updating the Javadoc of the `initializeMessages` to mention BOM, at all?

Let me know if something else should be addressed. @tjwatson @laeubi

tjwatson reviewed (commented):
While I would like to be able to auto-detect the encoding by reading the first 3 bytes, I am unsure about how reliable it is.

Why are we reluctant to add new API that lets the encoding be passed?  To me it seems that the developer of the bundle should know what encoding they are using such that we do not have to guess.

ShadelessFox commented on bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java:
```diff
@@ -402,6 +407,35 @@ static void log(int severity, String message, Exception e) {
+		switch (is.read()) {
+		case 0xEF: // EF BB BF
+				if (is.read() == 0xBB && is.read() == 0xBF)
+					return StandardCharsets.UTF_8;
```
BOMs are pretty reliable.

As discussed above https://github.com/eclipse-equinox/equinox/pull/709#issuecomment-2514843620, it's very unlikely that a properly formatted `.properties` file will start with the characters `ï»¿`. As for UTF-8, the sequence `EF BB BF` corresponds to BOM for UTF-8.

The table showing byte order marks (BOM) can be found here:
https://en.wikipedia.org/wiki/Byte_order_mark#Byte-order_marks_by_encoding

ShadelessFox commented:
> Why are we reluctant to add new API that lets the encoding be passed?

Does it make sense to have both mechanisms? By default, it will try to detect BOM and choose an appropriate charset based on it. Otherwise, the programmer may explicitly specify the desired charset, bypassing BOM detection. Is that right?

If you think having a way to pass the desired charset is still desired, I'll bring it back.

ShadelessFox commented on bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java:
```diff
@@ -347,7 +352,7 @@ static void load(final String bundleName, Class<?> clazz) {
 			try {
 				final MessagesProperties properties = new MessagesProperties(fields, bundleName, isAccessible);
-				properties.load(input);
+				properties.load(newBomReader(input));
```
Right, I wanted to be sure that change is required.

tjwatson commented:
> > Why are we reluctant to add new API that lets the encoding be passed?
> 
> Does it make sense to have both mechanisms? By default, it will try to detect BOM and choose an appropriate charset based on it. Otherwise, the programmer may explicitly specify the desired charset, bypassing BOM detection. Is that right?
> 
> If you think having a way to pass the desired charset is still desired, I'll bring it back.

My main point is that a new API that takes the desired encoding is going to be the most reliable and have zero chance of breaking existing bundles.  That is my preferred solution.

ShadelessFox commented:
> My main point is that a new API that takes the desired encoding is going to be the most reliable and have zero chance of breaking existing bundles. That is my preferred solution.

Does that mean the proposed solution, which relies on BOM detection, is not desired and should be removed from the PR, leaving us with an overloaded method that takes `Charset`? I just want to be extra sure I understand your intentions.

tjwatson commented:
> Does that mean the proposed solution, which relies on BOM detection, is not desired and should be removed from the PR, leaving us with an overloaded method that takes `Charset`? I just want to be extra sure I understand your intentions.

Yes, that is my suggestion.  Leave the existing method unchanged.  Add a new method that takes the charset.  That seems the most simple way to add the function without any impact to existing users.

ShadelessFox commented:
Adding an overload method will not have any benefits but pain when migrating to using UTF-8 in message bundles. 

Since not all characters 8859-1 are valid UTF-8, changing the encoding of a single message bundle will require re-encoding all its properties files in UTF-8. This will potentially break existing fragment bundles that provide localization to external bundles, and the opposite issue of using UTF-8 encoded properties files in an external bundle that hasn't been migrated to using UTF-8 yet.

I did a small research and noticed that the OSGI bundle's localization files (`OSGI-INF/l10n/bundle`) are actually loaded using [`PropertyResourceBundle`](https://docs.oracle.com/javase/9/docs/api/java/util/PropertyResourceBundle.html). Its Javadoc says:

> Constructing a PropertyResourceBundle instance from an InputStream requires that the input stream be encoded in UTF-8. By default, if a [MalformedInputException](https://docs.oracle.com/javase/9/docs/api/java/nio/charset/MalformedInputException.html) or an [UnmappableCharacterException](https://docs.oracle.com/javase/9/docs/api/java/nio/charset/UnmappableCharacterException.html) occurs on reading the input stream, then the PropertyResourceBundle instance resets to the state before the exception, re-reads the input stream in ISO-8859-1, and continues reading.

Therefore, I propose switching to using `PropertyResourceBundle` for `NLS` as well.

merks commented:
Reusing something from the JDK that already is geared to address the same problem sounds promising...

tjwatson commented:
> Therefore, I propose switching to using PropertyResourceBundle for NLS as well.

It is worth a try.  My minor concern is that it will have a noticeable impact on initial startup on very large installs of Eclipse (with many bundles) given that we will now need to construct many PropertyResourceBundle to load the messages only to get the values out of and then throw the objects away for GC.

Like I mentioned before, this was a measurable performance issue, but that was long ago on much older Java versions.

tjwatson commented:
I don't see the advantage of using the `PropertyResourceBundle` over directly using `java.util.Properties.load(Reader)`.

merks commented:
I've not done a deep dive, but if I understand correctly, the `PropertyResourceBundle` handles these documented cases with the different encodings:

![image](https://github.com/user-attachments/assets/aa9291a0-7d5a-4eed-b60f-fe48ecfe5f62)

But `Properties.load` does not... 

I don't recall all the history, but I thought the primary reason for the NLS design what to reduce memory footprint, not so much about load performance.  I'm not sure about that though...

tjwatson commented:
Curiosity got the best of me and now I am confused.  We already call `Properties.load` here:
https://github.com/eclipse-equinox/equinox/blob/master/bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java#L350

I was thinking we could simply change that to use a Reader.  But that isn't going to work.  Turns out in Java 11 that changed `PropertyResourceBundle.load(InputStream)` to do the UTF-8 stuff:

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/PropertyResourceBundle.html#%3Cinit%3E(java.io.InputStream)

In Java 8 that was not the case:

https://docs.oracle.com/javase/8/docs/api/java/util/PropertyResourceBundle.html#PropertyResourceBundle-java.io.InputStream-

So we can certainly change to using `PropertyResourceBundle.load` but that only helps if on Java 11 or higher.  Not really an issue for the vast majority of users.  Equinox framework does support Java 8 still, but I am fine with bundles wanting to run on Java 8 not having this ability to use UTF-8.

ShadelessFox commented:
> Not really an issue for the vast majority of users. Equinox framework does support Java 8 still, but I am fine with bundles wanting to run on Java 8 not having this ability to use UTF-8.

Should the Javadoc for `NLS#initializeMessages` document new behavior regardless of differences between Java versions?

ShadelessFox commented:
For now, I have updated the PR once again. I followed the most straightforward way that doesn't involve complex refactoring.

tjwatson commented:
> Should the Javadoc for `NLS#initializeMessages` document new behavior regardless of differences between Java versions?

That is probably a good idea, it allows us to claim a behavior enhancement to the API contract and have a minor version bump that bundles can depend on if they want.

tjwatson reviewed (approved):
LGTM

Thanks for you patience on this review!

merks commented on bundles/org.eclipse.osgi/supplement/src/org/eclipse/osgi/util/NLS.java:
```diff
@@ -138,6 +139,16 @@ public static String bind(String message, Object[] bindings) {
+	 * The properties files are read using the default encoding for resource bundles:
+	 * <ul>
+	 *   <li>In Java 8, resource files are read using ISO-8859-1 character encoding. Characters outside
+	 *   its range must be represented using Unicode escape sequences (e.g., <code>&#92;XXXX</code>).</li>
```
It's kind of a nit, but is anything other than `\uXXXX` supported in a *.properties file?   It seems slightly wrong to leave out the `u`.

ShadelessFox replied:
Ah, nice catch. I erased `u` by accident while escaping the backslash.

ShadelessFox replied:
Pushed the fix.

merks reviewed (commented):
It generally looks very good with tiny remark about the `\uXXXX` versus `\XXXX`.

merks reviewed (approved):
Thanks for being responsive to feedback, for writing tests, and for producing such a nice improvement that aligns better with the modern Java support! 🥇

ShadelessFox commented:
I truly appreciate all of you for sharing your feedback and guiding me along the way :^)

jukzi commented:
Please note that this PR is blamed to fail I-Build, see https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/issues/2648 @ShadelessFox @merks

merks commented:
Yes, it will be better if (and when) the PR build is as strict about Javadoc errors as the I-build.

ShadelessFox commented:
I'm sorry for causing havoc. I believe I had no way to guarantee it builds everywhere; I relied solely on the PR's checks.

merks commented:
No need to apologize.  The PR builds should have checked this given that the I-build is so strict about it later on.  Everything is back on track and we have a successful I-Build now.

akurtakov commented:
I can only say that from my POV work on enhancing our releng process is the most important now.
